### PR TITLE
Fix Relational Variables

### PR DESCRIPTION
**Pull request by APickledWalrus**
### Description
Reopening of #3582 (note: I have not touched anything, just reopened this for the reason of continued discussion)

This PR fixes the NotSerializableException when PersistentDataHolder is a registered ClassInfo. Relational Variables are enabled again too. This does not fix the actual issue with Yggdrasil, so it is more of a workaround. If anyone has any ideas for solutions, I am open to hearing them.

edit: thanks to tpgames for helping me out ;D

---
**Target Minecraft Versions:** 1.14+
**Requirements:** None
**Related Issues:** #3185

### CODE REVIEW

**Review by FranKusmiruk — APPROVED** [review]

**Comment by APickledWalrus** [comment]
notes to self:
- https://github.com/SkriptLang/Skript/blob/e4dc9ef722916dedb74590548b260fc8c84c09c9/src/main/java/ch/njol/skript/expressions/ExprRelationalVariable.java#L122 should just use the isList method of Variable
- when serializing a ByteArrayOutputStream should probably be used instead

**Comment by APickledWalrus** [comment]
A few things to look at:
- User friendliness of syntax
- Performance
- Missing syntax (e.g. does loop-index work) (I don't think it currently does, but it may not even be feasible)

File: src/main/java/ch/njol/skript/util/ListVariablePersistentDataType.java [hunk 1]
```diff
@@ -62,32 +65,34 @@
...
 
-			bb.putInt(indexBytes.length);
-			bb.put(indexBytes);
-
-			bb.putInt(typeBytes.length);
-			bb.put(typeBytes);
-
-			bb.putInt(entry.getValue().data.length);
-			bb.put(entry.getValue().data);
+			try {
```
Comment (Pikachu920) [inline src/main/java/ch/njol/skript/util/ListVariablePersistentDataType.java:75]: could you explain why the change from ByteBuffer to ByteArrayOutputSTream was required?

**Review by Pikachu920 — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/util/ListVariablePersistentDataType.java (hunk 1 above)
Comment (TPGamesNL) [reply src/main/java/ch/njol/skript/util/ListVariablePersistentDataType.java:75]: It wasn't required, but it'd be more convenient than having to pre-allocate the size of the buffer by calculating the length in bytes.
Also it may be better to use `DataOutputStream` as well to avoid having to use ByteBuffer just for the int writing

Comment (APickledWalrus) [reply src/main/java/ch/njol/skript/util/ListVariablePersistentDataType.java:75]: Sure, I'll take a look at making further improvements here. I would also like to add that I found better performance with this new method compared to the old method where it was manually calculated.

**Comment by TheLimeGlass** [comment]
Closing, see https://github.com/SkriptLang/Skript/pull/5158
